cmd/lotus-shed/indexes.go

```diff
@@ -483,87 +492,145 @@ var inspectEventsCmd = &cli.Command{
+		if err != nil {
+			return err
+		}
+		stmtSelectEventEntries, err := db.Prepare(selectEventEntries)
+		if err != nil {
+			return err
+		}
 
-		processHeight := func(ctx context.Context, ts *types.TipSet, receipts []*types.MessageReceipt) error {
-			tsKeyCid, err := ts.Key().Cid()
+		processHeight := func(ctx context.Context, messages []lapi.Message, receipts []*types.MessageReceipt) error {
+			tsKeyCid, err := currTs.Key().Cid()
 			if err != nil {
-				return fmt.Errorf("failed to get tipset key cid: %w", err)
+				return xerrors.Errorf("failed to get tipset key cid: %w", err)
 			}
 
-			var expectEvents int
-			var expectEntries int
+			var problems []string
+			var hasEvents bool
 
-			for _, receipt := range receipts {
-				if receipt.ExitCode != exitcode.Ok || receipt.EventsRoot == nil {
-					continue
+			checkEventAndEntryCounts := func() error {
+				// compare by counting events, using ChainGetEvents to load the events from the chain
+				expectEvents, expectEntries, err := chainEventAndEntryCountsAt(ctx, currTs, receipts, api)
+				if err != nil {
+					return err
 				}
-				events, err := api.ChainGetEvents(ctx, *receipt.EventsRoot)
+				if expectEvents > 0 {
+					hasEvents = true
+				}
+
+				actualEvents, actualEntries, err := dbEventAndEntryCountsAt(currTs, stmtEventCount, stmtEntryCount)
 				if err != nil {
-					return fmt.Errorf("failed to load events for tipset %s: %w", currTs, err)
+					return err
 				}
-				expectEvents += len(events)
-				for _, event := range events {
-					expectEntries += len(event.Entries)
+
+				if actualEvents != expectEvents {
+					problems = append(problems, fmt.Sprintf("expected %d events, got %d", expectEvents, actualEvents))
+				}
+				if actualEntries != expectEntries {
+					problems = append(problems, fmt.Sprintf("expected %d entries, got %d", expectEntries, actualEntries))
 				}
+
+				return nil
 			}
 
-			var problems []string
+			// Compare the AMT roots: we reconstruct the event AMT from the database data we have and
+			// compare it with the on-chain AMT root from the receipt. If it's the same CID then we have
+			// exactly the same event data. Any variation, in number of events, and even a single byte
+			// in event data, will be considered a mismatch.
+
+			// cache for address -> actorID because it's typical for tipsets to generate many events for
+			// the same actors so we can try and avoid too many StateLookupID calls
+			addrIdCache := make(map[address.Address]abi.ActorID)
+
+			eventIndex := 0
+			for msgIndex, receipt := range receipts {
+				if receipt.EventsRoot == nil {
+					continue
+				}
+
+				amtRoot, has, problem, err := amtRootForEvents(
+					ctx,
+					api,
+					tsKeyCid,
+					prevTs.Key(),
+					stmtSelectEventIdAndEmitter,
+					stmtSelectEventEntries,
+					messages[msgIndex],
+					addrIdCache,
+				)
+				if err != nil {
+					return err
+				}
+				if has && !hasEvents {
+					hasEvents = true
```

Comment:
Why can't this just be `hasEvents=has` without any if guards ?

Reply:
cause it's run for each message in the tipset, we just want to set this if it encounters at least one event in one of the messages